roxie/udplib/udptrr.cpp

```diff
@@ -270,71 +710,118 @@ class CReceiveManager : implements IReceiveManager, public CInterface
-                length++;
-            }
-            assert(prev == tail);
-            assert(lookfor==nullptr);
-            assert(numEntries==length);
-#endif
+            return (numPermits != 0);
         }
-    public:
-        unsigned length() const { return numEntries; }
-        operator UdpSenderEntry *() const
+
+        bool hasUnusedPermit() const
         {
-            return head;
+            return (numPermits != MAX_PERMITS);
         }
-        void append(UdpSenderEntry *sender)
+
+        inline SendPermit * queryPermit(unsigned flowSeq)
         {
-            if (tail)
+            for (unsigned i=0; i < MAX_PERMITS; i++)
             {
-                tail->nextSender = sender;
-                sender->prevSender = tail;
-                tail = sender;
+                if (permits[i].flowSeq == flowSeq)
+                    return &permits[i];
             }
-            else
+            return nullptr;
+        }
+
+        inline SendPermit & allocatePermit(unsigned permitTime, unsigned num)
+        {
+            for (unsigned i=0; i < MAX_PERMITS; i++)
             {
-                head = tail = sender;
+                SendPermit & permit = permits[i];
+                if (!permit.isActive())
+                {
+                    numPermits++;
+                    lastPermitTime = permitTime;
+
+                    permit.flowSeq = flowSeq;
+                    permit.sendSeq = sendSeq;
+                    permit.numPackets.store(num, std::memory_order_release);
+                    permit.permitTime = permitTime;
+                    return permit;
+                }
             }
-            numEntries++;
-            checkListIsValid(sender);
+            throwUnexpected();
         }
-        void remove(UdpSenderEntry *sender)
+
+        void revokePermit(SendPermit & permit)
         {
-            if (sender->prevSender)
-                sender->prevSender->nextSender = sender->nextSender;
-            else
-                head = sender->nextSender;
-            if (sender->nextSender)
-                sender->nextSender->prevSender = sender->prevSender;
-            else
-                tail = sender->prevSender;
-            sender->prevSender = nullptr;
-            sender->nextSender = nullptr;
-            numEntries--;
-            checkListIsValid(nullptr);
+            permit.revokePermit();
+            numPermits--;
+        }
+
+        inline void updateNumReserved(unsigned flowSeq, unsigned num)
+        {
+            for (unsigned i=0; i < MAX_PERMITS; i++)
+            {
+                if (permits[i].flowSeq == flowSeq)
+                {
+                    permits[i].numPackets.store(num, std::memory_order_release);
+                    return;
+                }
+            }
+        }
+        inline void decPermit(unsigned msgSeq)
+        {
+            //Although this is a bit more work than matching by flowSeq it shouldn't be too inefficient.
```

Comment:
Usually the permit will be the same one that we matched last time. Can we do a quick check to tell?

Reply:
This code feels painful to be executing per received message